enterprise/reporting/reporting.go

```diff
@@ -361,126 +377,116 @@ func (r *DefaultReporter) mainLoop(ctx context.Context, c types.SyncerConfig) {
 	r.getReportsQueryTime = r.stats.NewTaggedStat(StatReportingGetReportsQueryTime, stats.TimerType, tags)
 	r.requestLatency = r.stats.NewTaggedStat(StatReportingHttpReqLatency, stats.TimerType, tags)
-	reportingLag := r.stats.NewTaggedStat(
-		"reporting_metrics_lag_seconds", stats.GaugeType, stats.Tags{"client": c.Label},
-	)
 
 	var lastReportedAtTime atomic.Time
 	lastReportedAtTime.Store(time.Now())
 
-	var wg sync.WaitGroup
-	wg.Add(1)
-	defer wg.Wait()
-	go func() {
-		// for monitoring reports pileups
-		defer wg.Done()
-		for {
-			lag := time.Since(lastReportedAtTime.Load())
-			reportingLag.Gauge(lag.Seconds())
-			select {
-			case <-ctx.Done():
-				return
-			case <-time.After(2 * time.Minute):
-			}
-		}
-	}()
+	g, ctx := errgroup.WithContext(ctx)
 
-	for {
-		if ctx.Err() != nil {
-			r.log.Infof("stopping mainLoop for syncer %s : %s", c.Label, ctx.Err())
-			return
-		}
-		requestChan := make(chan struct{}, r.maxConcurrentRequests.Load())
-		loopStart := time.Now()
-		currentMin := time.Now().UTC().Unix() / 60
-
-		getReportsStart := time.Now()
-		reports, reportedAt, err := r.getReports(currentMin, c.ConnInfo)
-		getReportsTimer.Since(getReportsStart)
-		getReportsCount.Observe(float64(len(reports)))
-		if len(reports) == 0 {
-			if err == nil {
-				lastReportedAtTime.Store(loopStart)
-			} else {
-				r.log.Errorw("getting reports", "error", err)
-			}
-			select {
-			case <-ctx.Done():
+	g.Go(func() error {
+		return r.emitLagMetric(ctx, c, &lastReportedAtTime)
+	})
+
+	g.Go(func() error {
+		for {
+			if ctx.Err() != nil {
 				r.log.Infof("stopping mainLoop for syncer %s : %s", c.Label, ctx.Err())
-				return
-			case <-time.After(r.sleepInterval.Load()):
+				return ctx.Err()
 			}
-			continue
-		}
-
-		lastReportedAtTime.Store(time.Unix(reportedAt*60, 0))
-		getAggregatedReportsStart := time.Now()
-		metrics := r.getAggregatedReports(reports)
-		getAggregatedReportsTimer.Since(getAggregatedReportsStart)
-		getAggregatedReportsCount.Observe(float64(len(metrics)))
-
-		errGroup, errCtx := errgroup.WithContext(ctx)
-		// default to -1 to allow unlimited concurrency
-		errGroup.SetLimit(config.GetInt("Reporting.maxConcurrentRequests", -1))
-		for _, metric := range metrics {
-			if r.whActionsOnly && metric.SourceCategory != "warehouse" {
-				// if whActionsOnly is true, we only send reports for wh actions sources
-				// we silently drop all other reports
+			requestChan := make(chan struct{}, r.maxConcurrentRequests.Load())
+			loopStart := time.Now()
+			currentMin := time.Now().UTC().Unix() / 60
+
+			getReportsStart := time.Now()
+			reports, reportedAt, err := r.getReports(currentMin, c.ConnInfo)
+			getReportsTimer.Since(getReportsStart)
+			getReportsCount.Observe(float64(len(reports)))
+			if len(reports) == 0 {
+				if err == nil {
+					lastReportedAtTime.Store(loopStart)
+				} else {
+					r.log.Errorw("getting reports", "error", err)
```

Comment:
I am noticing this as an anti-pattern. 

the `err` value should be checked immediately and then the returns values should be checked. 

It is a convention that if err is not nil, other values should not be populated, thus reports will be empty. However, not checking the error hide a bug, where getReports returns a partially filled array with an error. In that case, error will not be reported.